src/uct/ib/mlx5/ud/ud_mlx5.c

```diff
@@ -519,9 +520,33 @@ uct_ud_mlx5_iface_poll_rx(uct_ud_mlx5_iface_t *iface, int is_async)
     len   = ntohl(cqe->byte_cnt);
     VALGRIND_MAKE_MEM_DEFINED(packet, len);
 
+    /*
+     * Take the packet type from CQE, because:
+     * 1. According to Annex17_RoCEv2 (A17.4.5.1):
+     * For UD, the Completion Queue Entry (CQE) includes remote address
+     * information (InfiniBand Specification Vol. 1 Rev 1.2.1 Section 11.4.2.1).
+     * For RoCEv2, the remote address information comprises the source L2
+     * Address and a flag that indicates if the received frame is an IPv4,
+     * IPv6 or RoCE packet.
+     *
+     * According to PRM, for responder UD/DC over RoCE sl represents RoCE
+     * packet type as:
+     * bit 3    : when set R-RoCE frame contains an UDP header otherwise not
+     * Bits[2:0]: L3_Header_Type, as defined below
+     *     - 0x0 : GRH - (RoCE v1.0)
+     *     - 0x1 : IPv6 - (RoCE v1.5/v2.0)
+     *     - 0x2 : IPv4 - (RoCE v1.5/v2.0)
+     *
+     * The service level is the most significant byte of cqe->flags_rqpn.
+     *
+     * Alternatively, this could be detected by examining the packet contents
+     * as is done for non-mlx5 transports (see
+     */
+    gid_len = (((ntohl(cqe->flags_rqpn) >> 24) & UCT_IB_CQE_SL_PKTYPE_MASK) == 0x2) ?
```

Comment:
seems this would be enough

```suggestion
    gid_len = ((cqe->flags_rqpn & UCT_IB_CQE_SL_PKTYPE_MASK) == 0x2) ?
```

Reply:
Won't that break on big-endian systems? I was looking at the mlx5 provider which uses this to get the sl:

https://github.com/linux-rdma/rdma-core/blob/master/providers/mlx5/cq.c#L320

I see other instances of byte-swapping in ud_mlx5.c so figured I should do the same because the CQE is in big-endian format. Just want to verify before I go ahead and make the change.

Reply:
Yes, and the byte we care about is the MSB of flags_rqpn so it is the LSB on little endian (why just masking the value without a byte-swap and bit-shift will work on LE) but on BE you will get LE MSB which is not what we want. Correct?

Reply:
Right.
maybe now, when `UCT_IB_CQE_SL_PKTYPE_MASK`  is used only for mlx5, we can move it to `src/uct/ib/mlx5/ib_mlx5.h`  and define like
`UCT_IB_MLX5_CQE_SL_PKTYPE_MASK = 0x7000000` 
and do this:
` gid_len = ((cqe->flags_rqpn & htonl(UCT_IB_CQE_SL_PKTYPE_MASK)) == 0x2) ` (can be an inline function like `uct_ib_mlx5_cqe_is_grh_present`)
similarly to what `uct_ib_mlx5_cqe_is_grh_present` does using `UCT_IB_MLX5_CQE_FLAG_L3_IN_DATA` and `UCT_IB_MLX5_CQE_FLAG_L3_IN_CQE`

note flags_rqpn is:
![image](https://github.com/user-attachments/assets/5ceee8b5-0852-47cd-ab0e-8fe0f1594e84)

Reply:
Right.
maybe now, when UCT_IB_CQE_SL_PKTYPE_MASK is used only for mlx5, we can move it to src/uct/ib/mlx5/ib_mlx5.h and define like
UCT_IB_MLX5_CQE_SL_PKTYPE_MASK = 0x7000000
and do this:
gid_len = ((cqe->flags_rqpn & htonl(UCT_IB_MLX5_CQE_SL_PKTYPE_MASK )) == 0x2) (can be an inline function like uct_ib_mlx5_cqe_is_grh_present)
similarly to what uct_ib_mlx5_cqe_is_grh_present does using UCT_IB_MLX5_CQE_FLAG_L3_IN_DATA and UCT_IB_MLX5_CQE_FLAG_L3_IN_CQE

Reply:
That won't quite work the same way because it is not a single bit. I changed it to avoid the << 24. What do you think of the new implementation. The alternative would be to treat it as a single bit and use 0x02000000 as the bit to detect IPv4. I can do that easily enough.

Reply:
The current implementation looks over complicated to me. 
IMO we should go with @brminich 's suggestion: Regardless on endianness, when we want to check if specific bits from a 32-bit value match expected value, a single & operation along with == should be enough. No need to add overheads of conversion and bit shifts.

Reply:
we still want to convert a mask (constat) i guess

Reply:
If the valid values are 0-3 then this is the minimum code that is needed. If they are 0-2 (which appears to be the case) I can simplify it more. Will do that.

Reply:
Ok. Now treating the roce packet types as flags with bit 24 being IPv6 and bit 25 being IPv4.